File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k){
        if(k>=num.length()) return "0";
        // if(k == 0){
        //     return num;
        // }
        stack<char>stk;
        stk.push(num[0]);
        int i;
        for(i=1; i<num.length() ; i++){
            while(!stk.empty() and stk.top()> num[i] and k){
                stk.pop();
                k--;
            }
            stk.push(num[i]);
        }
        while(!stk.empty() and k--){
            stk.pop();
        }
        string res="";
        while(!stk.empty()){
            res += stk.top();
            stk.pop();
        }
        
        bool p = true;
        for(int i= res.length()-1;i>=0; i--){
            if(p){
                if(res[i] == '0'){
                res.pop_back();
                }                    
                else{
                    p = false;
                }
            }
        }
        reverse(res.begin(),res.end());
        return res ==""? "0":res;
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        StringBuilder s = new StringBuilder();
        for(int i=0 ; i<num.length()  ; i++){
            while(!s.isEmpty() && s.charAt(s.length()-1) > num.charAt(i) && k>0){
                s.deleteCharAt(s.length()-1);
                k--;
            } 
            s.append(num.charAt(i));
        }
        while(!s.isEmpty() && k>0){
            s.deleteCharAt(s.length()-1);
            k--;
        }
        int index = 0;
        // if(s.length()>1)
        while(index<s.length() &&s.charAt(index) == '0'){
            index++;
        }
        return index == s.length() ?"0":s.substring(index);
    }
}